### PR TITLE
ARROW-13764: [C++] Support CountOptions in grouped count distinct

### PR DESCRIPTION
**Pull request by lidavidm**
This works by filtering the values (trading off time for memory usage).

### CODE REVIEW

File: docs/source/cpp/compute.rst [hunk 1]
```diff
@@ -289,36 +289,42 @@ The supported aggregation functions are as follows. All function names are
...
 * \(1) If null values are taken into account, by setting the
   :member:`ScalarAggregateOptions::skip_nulls` to false, then `Kleene logic`_
   logic is applied. The min_count option is not respected.
 
-* \(2) CountMode controls whether only non-null values are counted (the
-  default), only null values are counted, or all values are counted.
+* \(2) CountMode controls whether only non-null values are counted
+  (the default), only null values are counted, or all values are
+  counted. For hash_distinct, it instead controls whether null values
+  are emitted.
```
Comment (pitrou) [inline docs/source/cpp/compute.rst:327]: Also add that the setting doesn't apply to the grouping keys?

**Review by pitrou — COMMENTED** [review]
Thank you. A couple comments, but looks mostly good to me.

**Comment by pitrou** [comment]
> This is because we make a 2 GiB size allocation in GrouperFastImpl::GetUniques

Really? Why does this happen?

**Comment by lidavidm** [comment]
> > This is because we make a 2 GiB size allocation in GrouperFastImpl::GetUniques
> 
> Really? Why does this happen?

I was trying to figure out why, but didn't get very far in untangling the internals before I decided that filtering before grouping was pointless anyways. This doesn't happen the current implementation, I might poke at this further later though.

**Comment by lidavidm** [comment]
Thanks for the suggestions. I believe I've addressed everything.

**Review by pitrou — APPROVED** [review]
+1, thank you!

**Pull request closed without merging**